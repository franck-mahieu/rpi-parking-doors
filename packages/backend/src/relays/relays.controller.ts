import { Controller, Get, Param, Post } from '@nestjs/common';
import {
  ApiBasicAuth,
  ApiOperation,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';

import { Roles } from '../auth/roles.decorator';
import { RelaysService } from './relays.service';

@ApiBasicAuth()
@ApiTags('relays')
@Controller('relays')
export class RelaysController {
  constructor(private readonly relaysService: RelaysService) {}

  @ApiOperation({
    description: `Close the relay number received in param during 10 sec, then reopen it. 
    \n The caller should have, at least, user right to perform this endpoint`,
  })
  @ApiResponse({
    status: 201,
    description: 'Request was received and relay is closed for 10 sec',
  })
  @Roles('user')
  @Post('openClose/:relayNumber')
  activateRelayForSeconds(@Param('relayNumber') relayNumber: number): boolean {
    return this.relaysService.activateRelayForSeconds(relayNumber, 10);
  }

  @ApiOperation({
    description: `find the current relay state of relay number received in param . 
    \n The caller should have, at least, user right to perform this endpoint`,
  })
  @ApiResponse({
    status: 200,
    description: 'The current relay state (true or false)',
  })
  @Roles('user')
  @Get('state/:relayNumber')
  getRelayState(@Param('relayNumber') relayNumber: number): boolean {
    return this.relaysService.getRelayState(relayNumber);
  }
}
