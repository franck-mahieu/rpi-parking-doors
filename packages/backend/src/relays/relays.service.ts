import { Injectable } from '@nestjs/common';
import * as onoff from 'onoff';

const { Gpio } = onoff;
import { ConfigService } from '../common/config/config.service';

@Injectable()
export class RelaysService {
  relays;
  virtualValue;

  constructor(private readonly config: ConfigService) {
    this.setRelayValue(Gpio.accessible);
  }

  activateRelayForSeconds(relayNumber: number, seconds: number): boolean {
    try {
      this.relays[relayNumber].writeSync(this.config.get('STATE_ON'));
      setTimeout(() => {
        this.relays[relayNumber].writeSync(this.config.get('STATE_OFF'));
      }, seconds * 1000);
      return true;
    } catch (e) {
      return false;
    }
  }

  getRelayState(relayNumber: number): boolean {
    if (
      this.relays &&
      this.relays[relayNumber] &&
      this.relays[relayNumber].readSync &&
      this.relays[relayNumber].readSync() === this.config.get('STATE_ON')
    ) {
      return true;
    }
    return false;
  }

  setRelayValue(gpioAccessible: boolean) {
    const relaysFromConf = this.config.get('RELAYS').split(',');

    /** usefull to launch the project in local (for developpment) */
    if (gpioAccessible) {
      this.relays = relaysFromConf.map(
        relay => new Gpio(parseInt(relay), 'out'),
      );
    } else {
      this.virtualValue = relaysFromConf.map(() => 0);
      this.relays = relaysFromConf.map((relay, index) => ({
        writeSync: value => {
          this.virtualValue[index] = value;
          return value;
        },
        readSync: () => {
          return this.virtualValue[index];
        },
      }));
    }
  }
}
