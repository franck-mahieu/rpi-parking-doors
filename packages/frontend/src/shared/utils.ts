const env = process.env.NODE_ENV;
const reactAppDevApiUrl = process.env.REACT_APP_DEV_API_URL;

export const getBaseUrl = (
  envParam: string | undefined,
  reactAppApiUrlParam: string | undefined,
) => {
  return envParam === 'development' ? reactAppApiUrlParam : '';
};

export const fetchUtils = async (uri: string, method: string, body?: any) => {
  const response = await fetch(`${getBaseUrl(env, reactAppDevApiUrl)}${uri}`, {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    },
    credentials: env === 'development' ? 'include' : 'same-origin',
    body: body,
  });
  const result: any = await response.json();
  if (
    (response.status >= 200 && response.status < 300) ||
    (result?.changes && result.changes !== 0)
  ) {
    return result;
  }
  return Promise.reject(result);
};
