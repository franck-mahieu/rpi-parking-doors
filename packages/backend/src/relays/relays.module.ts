import { Module } from '@nestjs/common';
import { RelaysController } from './relays.controller';
import { RelaysService } from './relays.service';

@Module({
  imports: [],
  controllers: [RelaysController],
  providers: [RelaysService],
})
export class RelaysModule {}
