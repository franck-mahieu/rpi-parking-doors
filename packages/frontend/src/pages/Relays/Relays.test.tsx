import React from 'react';
import renderer from 'react-test-renderer';

import * as utils from '../../shared/utils';

jest.mock('../../shared/utils');

import relays from './Relays';

describe('Relays.tsx', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.restoreAllMocks();
    //@ts-ignore
    utils.fetchUtils.mockResolvedValue(true);
  });

  describe('Relays', () => {
    it('should render component correctly', () => {
      const component = renderer
        .create(<relays.Relays guid={'guid'} />)
        .toJSON();
      expect(component).toMatchSnapshot();
    });

    it('should call onButtonClick on click on two buttons', () => {
      const component = renderer.create(<relays.Relays guid={'guid'} />);
      jest.spyOn(relays, 'onButtonClick').mockResolvedValue();
      component.root.findByProps({ id: 'relayState0' }).props.onClick();
      component.root.findByProps({ id: 'relayState1' }).props.onClick();
      expect(relays.onButtonClick).toHaveBeenCalledTimes(2);
    });
  });

  describe('getRelayState', () => {
    it('should call fetchUtils and setRelayState with valid params', async () => {
      const setRelayStateMock = jest.fn();
      const result = await relays.getRelayState({
        relayNumber: 0,
        setRelayState: setRelayStateMock,
        guid: 'guid',
      });
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/relays/state/0?guid=guid',
        'GET',
      );
      expect(setRelayStateMock).toHaveBeenCalledWith(true);
      expect(result).toEqual(true);
    });
  });

  describe('openCloseRelay', () => {
    it('should call fetchUtils with valid params', async () => {
      await relays.openCloseRelay({
        relayNumber: 0,
        guid: 'guid',
      });
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/relays/openClose/0?guid=guid',
        'POST',
      );
    });
  });

  describe('getRelayStateUntilRelayIsOff', () => {
    it('should call fetchUtils with and setRelayState with valid params', async () => {
      const setRelayStateMock = jest.fn();
      relays.getRelayState = jest
        .fn()
        .mockResolvedValueOnce(true)
        .mockResolvedValueOnce(false);
      jest.useFakeTimers();

      await relays.getRelayStateUntilRelayIsOff({
        relayNumber: 0,
        setRelayState: setRelayStateMock,
        guid: 'guid',
      });
      jest.advanceTimersByTime(5000);
      expect(relays.getRelayState).toHaveBeenCalledTimes(2);
    });
  });

  describe('onButtonClick', () => {
    it('should call openCloseRelay, getRelayState, getRelayStateUntilRelayIsOff', async () => {
      const setRelayStateMock = jest.fn();
      jest.spyOn(relays, 'openCloseRelay').mockResolvedValue();
      jest.spyOn(relays, 'getRelayState').mockResolvedValue(true);
      jest.spyOn(relays, 'getRelayStateUntilRelayIsOff').mockReturnValue();
      jest.useFakeTimers();
      await relays.onButtonClick({
        relayNumber: 0,
        setRelayState: setRelayStateMock,
        guid: 'guid',
      });
      jest.advanceTimersByTime(5000);

      expect(relays.openCloseRelay).toHaveBeenCalled();
      expect(relays.getRelayState).toHaveBeenCalled();
      expect(relays.getRelayStateUntilRelayIsOff).toHaveBeenCalled();
    });
  });

  describe('showActiveRelay', () => {
    it('should return enable when relay state is true', async () => {
      const result = await relays.showActiveRelay(true);
      expect(result).toEqual('enable');
    });

    it('should return disable when relay state is false', async () => {
      const result = await relays.showActiveRelay(false);
      expect(result).toEqual('disable');
    });
  });
});
