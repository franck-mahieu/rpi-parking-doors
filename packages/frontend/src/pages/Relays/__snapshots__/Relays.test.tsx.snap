// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relays.tsx Relays should render component correctly 1`] = `
<div
  className="relays"
>
  <h4>
    Ouverture des portes
  </h4>
  <button
    className="waves-effect waves-light btn btn-large disable"
    id="relayState0"
    onClick={[Function]}
  >
    Porte R+1
  </button>
  <br />
  <button
    className="waves-effect waves-light btn btn-large disable "
    id="relayState1"
    onClick={[Function]}
  >
    Porte du rez de chauss√©e
  </button>
</div>
`;
